### PR TITLE
Allow multiple updates in a single conformance test

### PR DESCRIPTION
**Pull request by tgummerer**
# Description

Currently conformance tests are one and done, meaning they simulate one update, and that's it.  However it's often useful to simulate multiple steps, e.g. resources being created and later destroyed.  This PR adds support for just that, having multiple steps in each conformance test, that all happen in the same stack.  

Note that this will be much easier to review ignoring whitespace changes, since a lot of code indentation changed.

### CODE REVIEW

File: cmd/pulumi-test-language/tests.go [hunk 1]
```diff
@@ -28,19 +28,24 @@ import (
...
-
 	// updateOptions can be used to set the update options for the engine.
 	updateOptions engine.UpdateOptions
 }
 
+type languageTest struct {
+	// TODO: This should be a function so we don't have to load all providers in memory all the time.
+	providers []plugin.Provider
+
+	runs []testRun
```
Comment (tgummerer) [inline cmd/pulumi-test-language/tests.go:46]: Happy to take naming suggestions here, I struggled with naming here.

File: cmd/pulumi-test-language/tests.go [hunk 2]
```diff
@@ -28,19 +28,24 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-type languageTest struct {
+type testRun struct {
+	subDir string
```
Comment (Frassle) [inline cmd/pulumi-test-language/tests.go:32]: I think I would have made this implicit. It's always painful trying to keep data in sync between two places and named subdirs also don't let you see at a glance what order the test is going to run in. 

I'd remove subDir and just name the folders "0","1", etc. With a special case for when len(runs) == 1 that there is no integer named directory, the test is just at the top level.

File: cmd/pulumi-test-language/tests.go [hunk 3]
```diff
@@ -55,81 +60,134 @@ var languageTests = map[string]languageTest{
...
-			// Check that we have the target in the snapshot, but not the other resource.
-			stack := snap.Resources[0]
-			require.Equal(l, resource.RootStackType, stack.Type, "expected a stack resource")
-			provider := snap.Resources[1]
-			assert.Equal(l, "pulumi:providers:simple", provider.Type.String(), "expected simple provider")
-			target := snap.Resources[2]
-			require.Equal(l, "simple:index:Resource", target.Type.String(), "expected simple resource")
-			require.Equal(l, "target", target.URN.Name(), "expected target resource")
+	},
+	"l2-multiple-tests": {
```
Comment (Frassle) [inline cmd/pulumi-test-language/tests.go:160]: l2-destroy might be a better name for this

File: cmd/pulumi-test-language/interface.go [hunk 4]
```diff
@@ -730,199 +730,231 @@ func (eng *languageTestServer) RunLanguageTest(
...
-		return nil, fmt.Errorf("create source dir: %w", err)
-	}
-
-	// Find and copy the tests PCL code to the source dir
-	err = copyDirectory(languageTestdata, filepath.Join("testdata", req.Test), sourceDir)
-	if err != nil {
-		return nil, fmt.Errorf("copy source test data: %w", err)
-	}
+	var result LResult
+	for i, run := range test.runs {
```
Comment (Frassle) [inline cmd/pulumi-test-language/interface.go:734]: This multi-run behaviour needs to be tested, e.g. see l2resourcesimple_test.go. We'll want a l2destroy_test.go.

File: cmd/pulumi-test-language/tests.go (hunk 2 above)
Comment (tgummerer) [reply cmd/pulumi-test-language/tests.go]: Good idea :heavy_check_mark:

File: cmd/pulumi-test-language/tests.go (hunk 3 above)
Comment (tgummerer) [reply cmd/pulumi-test-language/tests.go]: Agreed, that's much better :heavy_check_mark:

File: cmd/pulumi-test-language/interface.go (hunk 4 above)
Comment (tgummerer) [reply cmd/pulumi-test-language/interface.go:735]: Added a test :heavy_check_mark:

File: cmd/pulumi-test-language/l2destroy_test.go [hunk 5]
```diff
@@ -0,0 +1,280 @@
+// Copyright 2016-2023, Pulumi Corporation.
```
Comment (Frassle) [inline cmd/pulumi-test-language/l2destroy_test.go:1]:
```suggestion
// Copyright 2016-2024, Pulumi Corporation.
```

File: cmd/pulumi-test-language/l2destroy_test.go [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+					"value": structpb.NewBoolValue(true),
+				},
+			},
+		})
+		if err != nil {
+			return nil, fmt.Errorf("could not register resource: %w", err)
+		}
+	}
+
+	h.currentRun++
```
Comment (Frassle) [inline cmd/pulumi-test-language/l2destroy_test.go:240]: Doesn't Run get called twice per update? Once for the engine to preview then again for the actual up? We should definitely assert what we expect currentRun to be here. I think it'll be 0,1,2,3 and the if statement above should be for currentRun == 0 or == 1.

File: cmd/pulumi-test-language/tests.go (hunk 1 above)
Comment (Frassle) [reply cmd/pulumi-test-language/tests.go:45]: runs seems fine to me

File: cmd/pulumi-test-language/tests.go [hunk 7]
```diff
@@ -55,81 +59,132 @@ var languageTests = map[string]languageTest{
...
+					provider := snap.Resources[1]
+					assert.Equal(l, "pulumi:providers:simple", provider.Type.String(), "expected simple provider")
+					// check that both expected resources are in the snapshot
+					simple := snap.Resources[2]
+					assert.Equal(l, "simple:index:Resource", simple.Type.String(), "expected simple resource")
+					simple2 := snap.Resources[3]
+					assert.Equal(l, "simple:index:Resource", simple2.Type.String(), "expected simple resource")
+				},
+			},
+			{
```
Comment (Frassle) [inline cmd/pulumi-test-language/tests.go:176]: For both these asserts can we check names as well

File: cmd/pulumi-test-language/l2destroy_test.go (hunk 6 above)
Comment (tgummerer) [reply cmd/pulumi-test-language/l2destroy_test.go:240]: No, we don't currently don't run previews in the conformance tests as far as I can see.  We're setting `SkipPreview` to true [here](https://github.com/pulumi/pulumi/blob/master/cmd/pulumi-test-language/interface.go#L880), and then only run [`stack.Update`](https://github.com/pulumi/pulumi/blob/master/cmd/pulumi-test-language/interface.go#L897), never `stack.Preview`.

Comment (Frassle) [reply cmd/pulumi-test-language/l2destroy_test.go:244]: Ok we should still assert that currentRun is only ever 0 or 1 then.

Comment (Frassle) [reply cmd/pulumi-test-language/l2destroy_test.go:244]: Probably at the very start of the function

**Review by Frassle — APPROVED** [review]

**Pull request merged**